packages/fxa-auth-server/lib/payments/paypal-processor.ts

```diff
@@ -0,0 +1,335 @@
+
+  /**
+   * Determine if the given invoice is within its grace period for
+   * additional payment retries or waiting on a pending transaction.
+   *
+   * Calculated by adding the `graceDays` in seconds to the invoice
+   * creation time. A `graceDays` value of 1 allows an invoice to be
+   * open for 24 hours hours (1 day) before not in grace period, etc.
+   *
+   * @param invoice
+   */
+  private inGracePeriod(invoice: Stripe.Invoice): boolean {
+    const graceDaysInSeconds = this.graceDays * 24 * 60 * 60;
+    return Date.now() / 1000 < invoice.created + graceDaysInSeconds;
+  }
+
+  private cancelInvoiceSubscription(invoice: Stripe.Invoice) {
+    // TODO: should send an email about subscription cancelled as overdue
+    return Promise.all([
+      this.stripeHelper.markUncollectible(invoice),
+      this.stripeHelper.cancelSubscription(invoice.subscription as string),
+    ]);
+  }
+
+  /**
+   * Ensures that an invoice matches the payment attempts returned from a PayPal
+   * Transaction Search.
+   *
+   * In general this should be the case, but in the event that an IPN notification
+   * is missed a mismatch could occur.
+   *
+   * @param invoice
+   * @param transactions
+   */
+  private async verifyAttemptcount(
+    invoice: Stripe.Invoice,
+    transactions: TransactionSearchResult[]
+  ) {
+    const paymentAttempts = this.stripeHelper.getPaymentAttempts(invoice);
+    if (transactions.length !== paymentAttempts) {
+      await this.stripeHelper.updatePaymentAttempts(
+        invoice,
+        transactions.length
+      );
+    }
+  }
+
+  /**
+   * Handles the condition where a transaction for an invoice has been paid.
+   *
+   * This occurrs when a previously pending transaction becomes paid, and the
+   * invoice should now be marked as paid.
+   *
+   * Returns whether the invoice has been paid.
+   *
+   * @param invoice
+   * @param transactions
+   */
+  private async handlePaidTransaction(
+    invoice: Stripe.Invoice,
+    transactions: TransactionSearchResult[]
+  ): Promise<boolean> {
+    const customer = invoice.customer;
+    const successTransactions = transactions.filter((t) =>
+      ['Completed', 'Processed'].includes(t.status)
+    );
+    if (successTransactions.length) {
+      if (successTransactions.length > 1) {
+        this.log.error('multipleCompletedTransactions', {
+          customer,
+          invoiceId: invoice.id,
+          transactionCount: successTransactions.length,
+          excessTransactions: successTransactions
+            .slice(1)
+            .map((t) => t.transactionId),
+        });
+        reportSentryError(
+          new Error('Multiple completed payments for invoice: ' + invoice.id)
+        );
+      }
+      const transaction = successTransactions[0];
+      await Promise.all([
+        this.stripeHelper.updateInvoiceWithPaypalTransactionId(
+          invoice,
+          transaction.transactionId
+        ),
+        this.stripeHelper.payInvoiceOutOfBand(invoice),
```

Comment:
How important is the order of these two calls?  In the (likely rare) scenario when one resolves but the other rejects, would that put an invoice in a state that we don't handle?

Reply:
If adding the transaction id works, but it isn't marked paid, then the next run through when it checks for completed transactions it will 'fix it. If its vice versa, then we end up with an invoice properly marked as closed but without a transaction id. Nice catch, I'm going to make these sequential so we record the transaction id first.